src/main/java/Matrices/ClusterAlgorithms/Thebeau.java

```diff
@@ -0,0 +1,304 @@
+            Grouping highestBidder = null;
+            Grouping secondHighestBidder = null;
+            double highestBid = -1;
+            double secondHighestBid = -1;
+            for (Map.Entry<Grouping, Double> entry : bids.entrySet()) {
+                if (Double.compare(entry.getValue(), highestBid) > 0) {
+                    secondHighestBidder = highestBidder;
+                    secondHighestBid = highestBid;
+
+                    highestBidder = entry.getKey();
+                    highestBid = entry.getValue();
+                } else if (Double.compare(entry.getValue(), secondHighestBid) > 0) {
+                    secondHighestBidder = entry.getKey();
+                    secondHighestBid = entry.getValue();
+                }
+            }
+
+            // choose a number between 0 and randBid to determine if it should make a suboptimal change
+            SymmetricDSMData tempMatrix = matrix.createCopy();
+            item = tempMatrix.getRows().get(n);  // update item to the item from the new matrix so that it is not modifying a copy
+            int nBid = generator.nextInt(randBid) + 1;  // add one to randBid because with truncation nBid will never be equal to randBid
+
+            Grouping oldGroup = item.getGroup1();
+            if (nBid == randBid) {  // assign item group to second-highest bidder
+                tempMatrix.setItemGroup(item, secondHighestBidder);
+
+            } else {  // assign to highest bidder
+                tempMatrix.setItemGroup(item, highestBidder);
+            }
+            deleteClusterIfEmpty(tempMatrix, oldGroup);
+
+            // choose a number between 0 and randAccept to determine if change is permanent regardless of it being optimal
+            int nAccept = generator.nextInt(randAccept) + 1;  // add one to randAccept because with truncation nAccept will never be equal to randAccept
+            Double newCoordinationScore = (Double) getCoordinationScore(tempMatrix, optimalSizeCluster, powcc, calculateByWeight).get("TotalCost");
+
+            if (nAccept == randAccept || newCoordinationScore < coordinationCost) {  // make the change permanent
+                coordinationCost = newCoordinationScore;
+                matrix = tempMatrix;
+
+                if (coordinationCost < bestSolutionCost) {  // save the new solution as the best one
+                    bestSolution = matrix.createCopy();  // use copy so this is permanent
+                    bestSolutionCost = coordinationCost;
+                }
+            }
+
+
+            String startTime = String.valueOf(Duration.between(absStart, start).toMillis());
+            String elapsedTime = String.valueOf(Duration.between(start, Instant.now()).toMillis());
+            debugString.append(i).append(",").append(startTime).append(",").append(elapsedTime).append(",").append(newCoordinationScore).append("\n");
+        }
+
+        if(debug) {
+            System.out.println(debugString);
```

Comment:
May be need use some logger, like Slf4j or Logger, they more useful and simple to use